### DISCUSSION TITLE
uhppoted-mqtt to AWS Greengrass aws.greengrass.clientdevices.mqtt.Moquette client

### DISCUSSION THREAD

**Discussion by irvintim** [uhppoted/uhppoted#14] [Q&A] [answered]
Hi,

First of all THANK YOU for creating this amazing suite of tools!

I'm having 2 problems getting this to work:

First, I'm struggling to get the uhppoted-mqtt to be able to connect to the AWS IoT Greengrass (v2) MQTT client (I'm using their Moquette lightweight client).

Here is the config file I am using (I had to gzip to attach for some reason *.conf isn't allowed):
[uhppoted.conf.gz](https://github.com/uhppoted/uhppoted/files/9174681/uhppoted.conf.gz)

I tried with "mqtt.connection.broker.certificate" commented out, and set to the 2 CA certs that are on the system.  But so far this is the error I keep getting:

> 2022/07/23 21:07:08 [client]   x509: certificate signed by unknown authority
> 2022/07/23 21:07:08 [client]   failed to connect to broker, trying next

Here is the certificate info for the cert on the local MQTT broker listening on localhost:8883:

> Certificate:
>     Data:
>         Version: 3 (0x2)
>         Serial Number:
>             e7:8e:74:38:70:d2:fa:79:f9:6b:87:85:9c:fe:a6:1e:f1:7a:db:63
>         Signature Algorithm: sha256WithRSAEncryption
>         Issuer: C = US, O = Amazon.com Inc., OU = Amazon Web Services, ST = Washington, L = Seattle, CN = Greengrass Core CA
>         Validity
>             Not Before: Jul 23 19:31:55 2022 GMT
>             Not After : Jul 30 19:31:55 2022 GMT
>         Subject: C = US, O = Amazon.com Inc., OU = Amazon Web Services, ST = Washington, L = Seattle, CN = aws.greengrass.clientdevices.mqtt.Moquette
> .....

Second, I don't seem to be able to restrict the directories that the app is using for working areas, so for instance when it receives and event I am getting this message:

> 2022/07/23 21:23:23 WARN  listen       rename /tmp/uhppoted1471835880.tmp /var/uhppoted/mqtt.events.retrieved: no such file or directory

But the command line I used to start the process is:

> /greengrass/v2/packages/artifacts/Office-Core-UhppotedComponent/1.0.7/uhppoted-mqtt -config /greengrass/v2/packages/artifacts/Office-Core-UhppotedComponent/1.0.7/uhppoted.conf -dir /greengrass/v2/work/Office-Core-UhppotedComponent -pid /greengrass/v2/work/Office-Core-UhppotedComponent/uhppoted-mqtt.pid -logfile /greengrass/v2/work/Office-Core-UhppotedComponent/uhppoted-mqtt.log run

I was hoping the "-dir" option would prevent it from trying to write in "/var/uhppoted" -- the app runs as the ggc_user instead of root, so it has limited access to directories.

This is the recipe I am using to start up the application (again sorry it's gzipped):
[recipe.gz](https://github.com/uhppoted/uhppoted/files/9174691/recipe.gz)

This is the heartbeat I'm getting in the logs:

> 2022/07/23 21:31:43 [client]   x509: certificate signed by unknown authority
> 2022/07/23 21:31:43 [client]   failed to connect to broker, trying next
> 2022/07/23 21:31:45 DEBUG watchdog
> 2022/07/23 21:31:45 WARN  watchdog     1 error
> 2022/07/23 21:31:50 DEBUG watchdog
> 2022/07/23 21:31:50 WARN  watchdog     1 error
> 2022/07/23 21:31:50 DEBUG health-check
> 2022/07/23 21:31:52 WARN  health-check 1 error

Here are the startup logs:

> 2022/07/23 20:53:35 START
> 2022/07/23 20:53:35 ERROR: open /etc/uhppoted/mqtt/client.cert: no such file or directory
> 2022/07/23 20:53:35 WARN Failed to get directory information for '/etc/uhppoted/mqtt/rsa/signing': stat /etc/uhppoted/mqtt/rsa/signing: no such file or directory
> 2022/07/23 20:53:35 WARN Failed to get directory information for '/etc/uhppoted/mqtt/rsa/encryption': stat /etc/uhppoted/mqtt/rsa/encryption: no such file or directory
> 2022/07/23 20:53:35 WARN  open /etc/uhppoted/mqtt/cards: no such file or directory
> 2022/07/23 20:53:35 INFO  mqttd        Listening on 0.0.0.0:60001
> 2022/07/23 20:53:35 INFO  mqttd        Publishing events to uhppoted/gateway/events
> 2022/07/23 20:53:35 INFO  listen       Initialising event listener
> 2022/07/23 20:53:35 INFO  listen       Listening

Also, one more thing I just noticed in the logs, I occasionally get this message, though it is listening as I see those being logged:

> 2022/07/23 21:41:10 WARN  UTC0311-L0x 123456789 incorrect listener address/port: 192.168.13.199:60001

Thanks for the help!

Tim

**Comment by uhppoted** [comment]
Hi Tim,

Glad you're finding it useful :-). Ok, let me address the easy ones:

1. _uhppoted-mqtt_ doesn't need to run as _root_ (and in fact shouldn't). It typically expects to run as _uhppoted:uhppoted_ but that's not mandatory and it can run as any user of your choice. The warning:
```
2022/07/23 21:23:23 WARN listen rename /tmp/uhppoted1471835880.tmp /var/uhppoted/mqtt.events.retrieved: no such file or directory
```
is because the default value for the `mqtt.events.index.filepath` configuration key is `/var/uhppoted/mqtt.events.retrieved`
```
...
# mqtt.events.index.filepath = /var/uhppoted/mqtt.events.retrieved
...
```
I can't remember exactly why it doesn't default to the _workdir_ - I'll make a note to fix it, but for the interim just change the value in the _uhppoted.conf_ file.

2. The listener address warning is from the system health-check:
```
2022/07/23 21:41:10 WARN UTC0311-L0x 123456789 incorrect listener address/port: 192.168.13.199:60001
```
In your _uhppoted.conf_ DEVICES section you have controller 123456789 defined:
```
...
UT0311-L0x.123456789.address = 192.168.13.120:60000
UT0311-L0x.123456789.door.1 = Front Door
...
```
The health check is reporting that the event listener address for `123456789` is not set correctly (not unexpectedly - `123456789` looks a bit bogus). Remove it from _uhppoted.conf_ (or replace it with the correct controller information) and the warning should go away. You may need to manually set the controller event listener address (using the CLI).

I don't have a ready answer for the certificate issues I'm afraid - am going to have to dig into it which is going to take a day or two. _Go_ has some particular requirements around TLS certificates and I suspect it's something to do with that but let me look into it. 

Something you *could* try in the interim is to add the AWS or GreenGrass root CA certificate to the system trust store (e.g. on [Ubuntu](https://ubuntu.com/server/docs/security-trust-store)) and then comment out the broker certificate entries in _uhppoted.conf_ - if _uhppoted-mqtt_ can't load the broker certificate it defaults to using the system trust store:

Ref. https://pkg.go.dev/crypto/tls#Config:
```
...
  // RootCAs defines the set of root certificate authorities
  // that clients use when verifying server certificates.
  // If RootCAs is nil, TLS uses the host's root CA set.
  RootCAs *[x509](https://pkg.go.dev/crypto/x509).[CertPool](https://pkg.go.dev/crypto/x509#CertPool)
...
```

TBH, I'm not particularly hopeful but it's something to try at least.

**Comment by uhppoted** [comment]
One other thing to try - the error:
```
2022/07/23 21:07:08 [client] x509: certificate signed by unknown authority
```
suggests that the root CA isn't in the system trust store and also not part of the broker certificate PEM chain. You could try concatenating the root CA PEM and the broker certificate PEM into a single file and using that as the broker certificate for _uhppoted-mqtt_.

The order in which the certificates are presented in the combined file is important - as far as I remember it should be broker certificate first followed by the CA certificate:

**Comment by irvintim** [comment]
Thanks for getting back to me so quickly:

1. File path in the config did the trick, thanks.
2. The "123456789" serial number was just me obfuscating that for this request, I'm not sure if having our serial # in the wild is a bad thing or not.  Our config has the actual serial number listed.
Also, I did use the CLI to set the listener address before starting up the mqtt process:
```
root@york-rpi:/greengrass/v2# uhppote-cli set-listener 123456789 192.168.13.199:60001
123456789 true
```
(My actual serial number was in the command I sent, not 123456789).
Looking more closely I seem to get this message when it tries the first "health-check":
```
2022/07/24 01:42:38 START
2022/07/24 01:42:38 ERROR: open /etc/uhppoted/mqtt/client.cert: no such file or directory
2022/07/24 01:42:38 WARN  open /etc/uhppoted/mqtt/cards: no such file or directory
2022/07/24 01:42:38 WARN Failed to get directory information for '/etc/uhppoted/mqtt/rsa/encryption': stat /etc/uhppoted/mqtt/rsa/encryption: no such file or directory
2022/07/24 01:42:38 WARN Failed to get directory information for '/etc/uhppoted/mqtt/rsa/signing': stat /etc/uhppoted/mqtt/rsa/signing: no such file or directory
2022/07/24 01:42:38 INFO  mqttd        Listening on 0.0.0.0:60001
2022/07/24 01:42:38 INFO  mqttd        Publishing events to uhppoted/gateway/events
2022/07/24 01:42:38 INFO  listen       Initialising event listener
2022/07/24 01:42:38 INFO  listen       Listening
2022/07/24 01:42:38 [client]   x509: certificate signed by unknown authority
2022/07/24 01:42:38 [client]   failed to connect to broker, trying next
2022/07/24 01:42:43 DEBUG watchdog
2022/07/24 01:42:43 INFO  watchdog     OK
2022/07/24 01:42:48 DEBUG watchdog
2022/07/24 01:42:48 INFO  watchdog     OK
2022/07/24 01:42:53 DEBUG watchdog
2022/07/24 01:42:53 INFO  watchdog     OK
2022/07/24 01:42:53 DEBUG health-check
2022/07/24 01:42:56 WARN  UTC0311-L0x 123456789 incorrect listener address/port: 192.168.13.199:60001
2022/07/24 01:42:56 WARN  health-check 1 error
2022/07/24 01:42:58 DEBUG watchdog
2022/07/24 01:42:58 WARN  watchdog     1 error
2022/07/24 01:43:03 DEBUG watchdog
2022/07/24 01:43:03 WARN  watchdog     1 error
2022/07/24 01:43:08 DEBUG watchdog
2022/07/24 01:43:08 WARN  watchdog     1 error
```
I get this every time I restart `uhppoted-mqtt`

3. Great ideas about concatenating the root cert to the device cert, and adding the root certs to the trust store.  I tried both and no joy, still getting the ` x509: certificate signed by unknown authority` error.  


Tim

**Comment by uhppoted** [comment]
Thanks for trying - there was just a tiny chance it might be that easy. I'm going to have to set up a _GreenGrass_ device and work through the configuration and see what the problem with the certificates actually is - it is going to take a day or so though.

re. health-check. Hmmm, that is a bit puzzling - it almost sounds like something else is changing the controller configuration. Can you maybe run the CLI _get-listener_ command to check that it is still set correctly when the error is showing in the log?

**Comment by irvintim** [comment]
Thanks for looking into it! I'm searching through docs to make sure I didn't miss something -- there are lots of moving parts in Greengrass/IoT and I'm still figuring it all out.

On the health-check I did as you asked:
```
2022/07/24 04:16:36 START
2022/07/24 04:16:36 ERROR: open /etc/uhppoted/mqtt/client.cert: no such file or directory
2022/07/24 04:16:36 WARN  open /etc/uhppoted/mqtt/cards: no such file or directory
2022/07/24 04:16:36 WARN Failed to get directory information for '/etc/uhppoted/mqtt/rsa/signing': stat /etc/uhppoted/mqtt/rsa/signing: no such file or directory
2022/07/24 04:16:36 INFO  mqttd        Listening on 0.0.0.0:60001
2022/07/24 04:16:36 INFO  mqttd        Publishing events to uhppoted/gateway/events
2022/07/24 04:16:36 INFO  listen       Initialising event listener
2022/07/24 04:16:36 WARN Failed to get directory information for '/etc/uhppoted/mqtt/rsa/encryption': stat /etc/uhppoted/mqtt/rsa/encryption: no such file or directory
2022/07/24 04:16:36 INFO  listen       Listening
2022/07/24 04:16:36 [client]   EOF
2022/07/24 04:16:36 [client]   failed to connect to broker, trying next
2022/07/24 04:16:41 DEBUG watchdog
2022/07/24 04:16:41 INFO  watchdog     OK
2022/07/24 04:16:46 DEBUG watchdog
2022/07/24 04:16:46 INFO  watchdog     OK
2022/07/24 04:16:51 DEBUG watchdog
2022/07/24 04:16:51 DEBUG health-check
2022/07/24 04:16:51 INFO  watchdog     OK
2022/07/24 04:16:54 WARN  UTC0311-L0x 123456789  incorrect listener address/port: 192.168.13.199:60001
2022/07/24 04:16:54 WARN  health-check 1 error
2022/07/24 04:16:56 DEBUG watchdog
2022/07/24 04:16:56 WARN  watchdog     1 error
^C
root@york-rpi:/greengrass/v2/work/Office-Core-UhppotedComponent# uhppote-cli get-listener 123456789
123456789  192.168.13.199:60001
```

**Comment by uhppoted** [comment]
Ok - getting there slowly. I can now at least connect to the _Moquette_ MQTT broker. It seems that by default _Moquette_ has mutual TLS authentication enabled and the client is therefore required to present a valid client certificate:

- the broker CA certificate should be the _Greengrass Core Device CA_ which is located in the `/greengrass/v2/work/aws.greengrass.clientdevices.Auth/ca.pem` file
- the client certificate is located in the `/greengrass/v2/thingCert.crt` file
- the client key is located in the `/greengrass/v2/privKey.key` file
- the `/greengrass/v3/rootCA.pem` is Amazon's root CA cert and is seemingly no use to man or beast (at least in this particular scenario)

After setting up the _uhppoted.conf_ file with these certificates I can at least connect to the MQTT broker:
```
...
mqtt.connection.client.ID = <your AWS IoT thing name>
mqtt.connection.broker.certificate = /usr/local/etc/com.github.uhppoted/mqtt/greengrass/ca.pem
mqtt.connection.client.certificate = /usr/local/etc/com.github.uhppoted/mqtt/greengrass/thingCert.crt
mqtt.connection.client.key = /usr/local/etc/com.github.uhppoted/mqtt/privKey.key
...
```

However the _greengrass.log_ is now showing an auth error - which may be because I haven't yet figured out how to authorise the device. You weren't kidding about there being a LOT of moving parts!
```
2022-07-24T05:52:46.991Z [WARN] (nioEventLoopGroup-3-2) com.aws.greengrass.mqtt.moquette.ClientDeviceAuthorizer: Device isn't authorized to connect. {clientId=uhppoted-1822e4f73af, sessionId=2b25c09f-34b2-4729-9e6e-8e915c3c46d6}
2022-07-24T05:52:46.991Z [INFO] (nioEventLoopGroup-3-2) io.moquette.broker.MQTTConnection: Authenticator has rejected the MQTT credentials CId=uhppoted-1822e4f73af, username=-----BEGIN CERTIFICATE-----
```

re. health-check. Thanks for checking - that is weird! I'll go through the code and see what could possibly be happening to cause it.

**Comment by irvintim** [comment]
I tried using the client certs as you suggested, but still getting the `x509: certificate signed by unknown authority` message.

I then found this page: https://docs.aws.amazon.com/iot/latest/developerguide/x509-client-certs.html which seemed promising.  I created new client certificates and gave them the policies suggested on this page: https://docs.aws.amazon.com/greengrass/v2/developerguide/connect-client-devices.html#configure-iot-policy-requirement

Still no joy.

Then I found this page: https://docs.aws.amazon.com/iot/latest/developerguide/server-authentication.html and thought I had struck paydirt.  I added all the root certs listed on this page to my trust store, but still the same issue.

So frustrating.  Thanks for helping with this!

**Comment by uhppoted** [comment, marked as answer]
Hi,

Tracked down the health-check issue  - the health check is comparing the configured system listener address with the controller listener address and with the default listen-all address of `0.0.0.0` that is obviously not going to work. I've created an [issue](https://github.com/uhppoted/uhppoted-lib/issues/2) to fix it and hopefully it should be fixed in the next couple of days. For now you can either ignore it (recommended) or change your _listener_ address in _uhppoted.conf_ to `192.168.13.199` if it is too annoying.

With much tearing out of hair (and openssl), I've been able to get _uhppoted-mqtt_ configured to connect to _Moquette_ starting with a clean Ubuntu system. Part of the problem is that AWS IoT expects the 'things' to use Cloud Discovery and the workaround for distributing the broker certificate is undocumented:
- [Add docs about manual connection of client devices to GG Core without cloud discovery](https://github.com/awsdocs/aws-iot-greengrass-v2-developer-guide/issues/20)

Basically the steps I followed were:
- Create and install an _AWS GreenGrass core device_ with the _Auth_ and _Moquette_ components and then copy the _ca.pem_ certificate from `/greengrass/v2/work/aws.greengrass.clientdevices.Auth/ca.pem` to use as the _uhppoted-mqtt_ broker certificate.
- Create an AWS IoT _Thing_ for _uhppoted-mqtt_ and keep the generated key for use as the _uhppoted-mqtt_ client key. Once the `Thing`has been created, download the certificate from the _uhppoted-mqtt_ thing certificate page to use as the client certificate.
- In _uhppoted.conf_, set the `mqtt.client-id` to the name of the AWS Thing.

That seems to get _uhppoted-mqtt_ successfully connected to Moquette without any auth issues:
```
2022/07/24 18:43:06 uhppoted-mqtt service v0.8.0 - Linux (PID 44489)
2022/07/24 18:43:06 START
2022/07/24 18:43:06 INFO  mqttd        Listening on 127.0.0.1:60001
2022/07/24 18:43:06 INFO  mqttd        Publishing events to uhppoted/gateway/events
2022/07/24 18:43:06 INFO  listen       Fetching unretrieved events for device ID 405419896
2022/07/24 18:43:06 INFO  listen       No unretrieved events for device ID 405419896
2022/07/24 18:43:06 INFO  listen       Initialising event listener
2022/07/24 18:43:06 INFO  listen       Listening
2022/07/24 18:43:06 INFO  mqttd        Connected to tls://localhost:8883
2022/07/24 18:43:06 INFO  mqttd        Subscribed to uhppoted/gateway/requests
```

I still have to figure out how to route messages so that they show up in the AWS MQTT test client - that's going to be another couple of hours of wading through the labyrinth of documentation though.

I did all the above manually from the AWS GreenGrass console (i.e. not using a recipe) which was ploddy, time-consuming and error-prone so not recommended but at least it gets to a basic working system.

OpenSSL was super-useful in working through all the issues:
```
1. To see Moquette's certificate chain:
    openssl s_client -connect localhost:8883 -showcerts

2. To connect with client auth:
    openssl s_client -CAfile AmazonRootCA1.pem -cert client.cert -key client.key -connect localhost:8883 
    openssl s_client -CAfile ca.pem -cert client.cert -key client.key -connect localhost:8883 

```

**Comment by irvintim** [comment]
You are amazing thank you for your help here. I'll try it out this
afternoon.

I was contemplating using your library to create a Python module that used
the Amazon SDK to put messages into core device's MQTT. But if this works,
this should be a lot easier.

I'm curious how it works to have a separate "thing" without it being a
separate device. Sounds like my project for the afternoon is well defined.

Tim

**Comment by uhppoted** [comment]
lol .. there was definitely a point where I was also thinking "_..maybe the way to go is Python and the DLL.._". 

It's certainly an alternative because _uhppoted-mqtt_ was originally designed with the intention of using it as a gateway for mobile apps so it has a lot of functionality that may be superfluous. NodeRED also has an MQTT node so that's another alternative, depending on how you feel about NodeJS.

Anyway, the good news is that events and messages are now viewable in the AWS console MQTT test client - the critical bit was that the bridge need to be configured to route to _IotCore_:
```
{
	"mqttTopicMapping": {
		"ClientDeviceEvents": {
			"topic": "uhppoted/#",
			"targetTopicPrefix": "vultr/",
			"source": "LocalMqtt",
			"target": "IotCore"
		}
	}
}
```

Sample event from the AWS console:
```
{
  "message": {
    "event": {
      "event": {
        "access-granted": false,
        "card-number": 8165538,
        "device-id": 405419896,
        "direction": {
          "code": 1,
          "description": "in"
        },
        "door-id": 3,
        "event-id": 85,
        "event-reason": {
          "code": 6,
          "description": "no access rights"
        },
        "event-type": {
          "code": 1,
          "description": "card swipe"
        },
        "timestamp": "2022-07-24 11:31:55 UTC"
      }
    }
  },
  "hmac": "330f88a2eccfc8f9403f28ccc975a8478adca47b5f212e53d6166941cd38108a"
}
```

Good luck with your afternoon project :-) ... let me know how it goes!

**Comment by irvintim** [comment]
Success!!!

I needed to install the Auth and Bridge components, and then followed your suggestions and voila:

```
vultr/uhppoted/gateway/events
July 24, 2022, 22:39:50 (UTC-0700)
{
  "message": {
    "event": {
      "event": {
        "access-granted": true,
        "card-number": XXXXXXXX,
        "device-id": 123456789,
        "direction": 1,
        "door-id": 1,
        "event-id": 152,
        "event-reason": 44,
        "event-type": 2,
        "timestamp": "2022-07-24 22:39:49 PDT"
      }
    }
  },
  "hmac": "8c652ee97914731d75274888471055ec1bce3f3b4eeb38333f6df98a62beb640"
}
```

I updated the listener address to avoid the watchdog and health check errors, I'll revert it back when that bug is squashed.

I can't thank you enough for your help getting this going.

Tim

**Comment by uhppoted** [comment]
Oh! That's **great** news!! Nice!!

re. help. Pleasure .. \<grin\> TBH, it gave me the necessary incentive to actually dive into the _AWS Greengrass_ documentation which I knew had to be done but had been avoiding. I'll get around to writing this up as a cookbook type thing for future use because it's bound to come up again.

Have also had a brief look at the [_Greengrass Discovery_](https://docs.aws.amazon.com/greengrass/v2/developerguide/greengrass-discover-api.html) service and it's very AWS specific so not really suitable for including in _uhppoted-httpd_ - but it does look like it could be used via e.g a bash script, which might simplify certificate configuration. Will experiment a bit..

I'll get on to fixing the health check and post an update here when it's ready - it just needs a little thinking about first.

**Comment by uhppoted** [comment]
Just a quick update to let you know a provisional fix for the health check [issue](https://github.com/uhppoted/uhppoted-lib/issues/2) is in the master branch of [uhppoted-mqtt](https://github.com/uhppoted/uhppoted-mqtt). At time of writing it has only been tested on _MacOS_ and _Ubuntu_.

You'll have to build it if you don't want to wait for the next release (probably a week or two) but building it is straightforward - clone the repo and run `make build`.

**Reply by irvintim** [reply]
Sorry that it took me a while to get back to this, I built `uhppoted-mqtt` on the Raspberry PI I'm using as the dev Greengrass Core/Thing.  The good news is that the problem with the `0.0.0.0` listen address and the healthcheck has been resolved, however the bad news is that it no longer works when I specify the interface address specifically.

**Works with 0.0.0.0:60001:**
```
2022/08/02 18:00:49 uhppoted-mqtt service v0.8.1 - Linux (PID 129681)
2022/08/03 01:00:49 START
2022/08/02 18:00:49 WARN  open /etc/uhppoted/mqtt/rsa/signing/mqttd.key: no such file or directory
2022/08/02 18:00:49 WARN  open /etc/uhppoted/mqtt/rsa/encryption/mqttd.key: no such file or directory
2022/08/02 18:00:49 WARN  stat /etc/uhppoted/mqtt/rsa/signing: no such file or directory
2022/08/02 18:00:49 WARN  stat /etc/uhppoted/mqtt/rsa/encryption: no such file or directory
2022/08/02 18:00:49 WARN  open /var/uhppoted/mqtt.nonce: no such file or directory
2022/08/02 18:00:49 WARN  open /var/uhppoted/mqtt.nonce.counters: no such file or directory
2022/08/03 01:00:49 WARN  open /etc/uhppoted/mqtt/cards: no such file or directory
2022/08/03 01:00:49 WARN Failed to get directory information for '/etc/uhppoted/mqtt/rsa/encryption': stat /etc/uhppoted/mqtt/rsa/encryption: no such file or directory
2022/08/03 01:00:49 INFO  mqttd        Listening on 0.0.0.0:60001
2022/08/03 01:00:49 INFO  mqttd        Publishing events to uhppoted/gateway/events
2022/08/03 01:00:49 WARN Failed to get directory information for '/etc/uhppoted/mqtt/rsa/signing': stat /etc/uhppoted/mqtt/rsa/signing: no such file or directory
2022/08/03 01:00:49 INFO  listen       Fetching unretrieved events for device ID 123456789
2022/08/03 01:00:49 INFO  listen       No unretrieved events for device ID 123456789
2022/08/03 01:00:49 INFO  listen       Initialising event listener
2022/08/03 01:00:49 INFO  listen       Listening
2022/08/03 01:00:50 INFO  mqttd        Connected to tls://172.19.0.1:8883
2022/08/03 01:00:50 INFO  mqttd        Subscribed to uhppoted/gateway/requests
2022/08/03 01:00:50 [store]    memorystore del: message 1 not found
2022/08/03 01:00:54 DEBUG watchdog            
2022/08/03 01:00:54 INFO  watchdog     OK
2022/08/03 01:00:59 DEBUG watchdog            
2022/08/03 01:00:59 INFO  watchdog     OK
2022/08/03 01:01:04 DEBUG watchdog            
2022/08/03 01:01:04 INFO  watchdog     OK
2022/08/03 01:01:04 DEBUG health-check        
2022/08/03 01:01:04 INFO  health-check refreshing interface IP address list
2022/08/03 01:01:07 INFO  health-check OK
2022/08/03 01:01:09 DEBUG watchdog            
2022/08/03 01:01:09 INFO  watchdog     OK
```

**Does not work with 192.168.13.199:60001:**
```
2022/08/02 18:03:20 uhppoted-mqtt service v0.8.1 - Linux (PID 129754)
2022/08/03 01:03:20 START
2022/08/02 18:03:20 WARN  open /etc/uhppoted/mqtt/rsa/signing/mqttd.key: no such file or directory
2022/08/02 18:03:20 WARN  open /etc/uhppoted/mqtt/rsa/encryption/mqttd.key: no such file or directory
2022/08/02 18:03:20 WARN  stat /etc/uhppoted/mqtt/rsa/signing: no such file or directory
2022/08/02 18:03:20 WARN  stat /etc/uhppoted/mqtt/rsa/encryption: no such file or directory
2022/08/03 01:03:20 WARN Failed to get directory information for '/etc/uhppoted/mqtt/rsa/signing': stat /etc/uhppoted/mqtt/rsa/signing: no such file or directory
2022/08/02 18:03:20 WARN  open /var/uhppoted/mqtt.nonce: no such file or directory
2022/08/02 18:03:20 WARN  open /var/uhppoted/mqtt.nonce.counters: no such file or directory
2022/08/03 01:03:20 WARN  open /etc/uhppoted/mqtt/cards: no such file or directory
2022/08/03 01:03:20 WARN Failed to get directory information for '/etc/uhppoted/mqtt/rsa/encryption': stat /etc/uhppoted/mqtt/rsa/encryption: no such file or directory
2022/08/03 01:03:20 INFO  mqttd        Listening on 192.168.13.199:60001
2022/08/03 01:03:20 INFO  mqttd        Publishing events to uhppoted/gateway/events
2022/08/03 01:03:20 INFO  listen       Fetching unretrieved events for device ID 123456789
2022/08/03 01:03:20 INFO  listen       No unretrieved events for device ID 123456789
2022/08/03 01:03:20 INFO  listen       Initialising event listener
2022/08/03 01:03:20 INFO  listen       Listening
2022/08/03 01:03:20 INFO  mqttd        Connected to tls://172.19.0.1:8883
2022/08/03 01:03:20 INFO  mqttd        Subscribed to uhppoted/gateway/requests
2022/08/03 01:03:20 [store]    memorystore del: message 1 not found
2022/08/03 01:03:25 DEBUG watchdog            
2022/08/03 01:03:25 INFO  watchdog     OK
2022/08/03 01:03:30 DEBUG watchdog            
2022/08/03 01:03:30 INFO  watchdog     OK
2022/08/03 01:03:35 DEBUG watchdog            
2022/08/03 01:03:35 INFO  watchdog     OK
2022/08/03 01:03:35 DEBUG health-check        
2022/08/03 01:03:35 INFO  health-check refreshing interface IP address list
2022/08/03 01:03:37 WARN  UTC0311-L0x 123456789  incorrect listener address/port: 192.168.13.199:60001
2022/08/03 01:03:37 WARN  health-check 1 error
2022/08/03 01:03:40 DEBUG watchdog            
2022/08/03 01:03:40 WARN  watchdog     1 error
```
**IP Addresses on host:**
```
root@york-rpi:/greengrass/v2/work/Office-Core-UhppotedComponent# hostname -I
192.168.13.199 172.19.0.1 172.18.0.1 172.17.0.1 169.254.161.110 169.254.154.55 169.254.130.245 2601:645:***********
```

**Comment by uhppoted** [comment]
Hi Tim,

That's just plain weird because I definitely tested it after implementing it. 

The problem is related to [net: UDPAddr's (and possibly TCPAddr's) AddrPort returns addresses with the wrong IP version](https://github.com/golang/go/issues/53607) so there may have been a change in the underlying `netip` package implementation because I did update to Go version to 1.18.3 just before releasing v0.8.1. Or it may have just slipped through the cracks somehow, in which case, my humble apologies.

I've pushed a fix to the `uhppoted-mqtt` repo if you want to try it out (Ref. https://github.com/uhppoted/uhppoted-lib/issues/2#issuecomment-1204299035)

**Reply by irvintim** [reply]
That fixed it.  Thanks!

BTW I took your advise and create an init script to pull down the right certs from AWS IoT before stating up, it's a pretty ugly shell script that does a number of preliminary things.  I'll send you a PR with it to see if you find it useful

**Comment by uhppoted** [comment]
Oh great - thanks for checking and confirming it!

re. script. Your script looks fine - thanks very much! It was on the TODO list but it was going to be a while before I got around to it so am very happy to have something.
